components/modules/ble.c

```diff
@@ -0,0 +1,1062 @@
+      }
+
+      // Now call struct.pack
+      // The arguments are the format string, and then the values
+      lua_rawgeti(L, LUA_REGISTRYINDEX, struct_pack_index);
+      lua_pushvalue(L, -3);   // dup the format
+      int nv = lua_rawlen(L, -3);
+      for (int i = 1; i <= nv; i++) {
+        lua_rawgeti(L, -2 - i, i);
+      }
+      if (lua_pcall(L, nv + 1, 1, 0)) {
+        PERROR();
+        message.errcode = BLE_ATT_ERR_UNLIKELY;
+        lua_pop(L, 2);
+        goto cleanup;
+      }
+      lua_remove(L, -2);   // remove the old value
+      // now have string (-1), struct(-2), chrtable (-3)
+    }
+    size_t datalen;
+    const char *data = lua_tolstring(L, -1, &datalen);
+
+    if (data) {
+      message.buffer = malloc(datalen);
+      if (message.buffer) {
+	message.length = datalen;
+	memcpy(message.buffer, data, datalen);
+	message.errcode = 0;
+      }
+    }
+    lua_pop(L, 1);
+    message.errcode = 0;
+  }
+  if (task_block->ctxt->op == BLE_GATT_ACCESS_OP_WRITE_CHR) {
+    // Push the value
+    lua_pushlstring(L, task_block->buffer, task_block->length);
+
+    // value, struct, chrtable
+    // If we have a struct, then unpack the values
+    if (!lua_isnoneornil(L, -2)) {
+      lua_createtable(L, 0, 0);
+      int stack_size = lua_gettop(L);
+      lua_rawgeti(L, LUA_REGISTRYINDEX, struct_unpack_index);
+      lua_pushvalue(L, -4);   // dup the format
+      lua_pushvalue(L, -4);   // dup the string
+      if (lua_pcall(L, 2, LUA_MULTRET, 0)) {
+        PERROR();
+        message.errcode = BLE_ATT_ERR_UNLIKELY;
+        lua_pop(L, 2);
+        goto cleanup;
+      }
+      int vals = lua_gettop(L) - stack_size;
+
+      // Note that the last entry is actually the string offset
+      for (int i = 1; i < vals; i++) {
+        lua_pushvalue(L, -(vals - i + 1));
+        lua_rawseti(L, -(vals + 2), i);
+      }
+      lua_pop(L, vals);
+      lua_remove(L, -2);
+      // Now have table, struct, chrtable
+    }
+    // If the value is a table of a single value, then
+    // treat as value
+    if (lua_istable(L, -1) && lua_rawlen(L, -1) == 1) {
+      lua_rawgeti(L, -1, 1);
+      lua_remove(L, -2);	// and throw away the table
+    }
+    // value, struct, chr table
+    // save to `value` if present
+    lua_getfield(L, -3, "value");
+    if (!lua_isnoneornil(L, -1)) {
+      lua_pop(L, 1);
+      lua_pushvalue(L, -1);
+      lua_setfield(L, -4, "value");
+    } else {
+      lua_pop(L, 1);
+    }
+    lua_getfield(L, -3, "write");
+    if (!lua_isnoneornil(L, -1)) {
+      lua_pushvalue(L, -4);   // the characterstics table
+      lua_pushvalue(L, -3);   // the value
+      if (lua_pcall(L, 2, 0, 0)) {
+        PERROR();
+	message.errcode = BLE_ATT_ERR_UNLIKELY;
+	lua_pop(L, 2);
+	goto cleanup;
+      }
+    } else {
+      lua_pop(L, 1);  // Throw away the null write pointer
+    }
+    lua_pop(L, 1);	// THrow away the value
+    message.errcode = 0;
+  }
+
+cleanup:
+  lua_pop(L, 2);
```

Comment:
I'm going to take your word for there always being 2 items on the stack here 😅
Personally I would've gone a gettop/settop to cover the just-in-case scenarios, but that' just me. Oh, and there's a case typo in the previous comment - `THrow` :)

Reply:
Yes. I will fix.